aptos-move/move-examples/token_objects/sources/token.move

```diff
@@ -0,0 +1,384 @@
+/// This defines an object-based Token. The key differentiating features from the Aptos standard
+/// token are:
+/// * Decouple token ownership from token data.
+/// * Explicit data model for token metadata via adjacent resources
+/// * Extensible framework for tokens
+///
+/// TODO:
+/// * Provide functions for mutability -- the capability model seems to heavy for mutations, so
+///   probably keep the existing model
+/// * Consider adding an optional source name if name is mutated, since the objects address depends
+///   on the name...
+/// * Update ObjectId to be an acceptable param to move
+module token_objects::token {
+    use std::error;
+    use std::option::{Self, Option};
+    use std::string::{Self, String};
+    use std::signer;
+    use std::vector;
+
+    // The token does not exist
+    const ETOKEN_DOES_NOT_EXIST: u64 = 0;
+    /// The provided signer is not the creator
+    const ENOT_CREATOR: u64 = 1;
+    /// Attempted to mutate an immutable field
+    const EFIELD_NOT_MUTABLE: u64 = 2;
+
+    use aptos_framework::object::{Self, CreatorRef, ObjectId};
+
+    use token_objects::collection::{Self, Royalty};
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// Represents the common fields to all tokens.
+    struct Token has key {
```

Comment:
how do you plan to support on-chain traits?

Reply:
See `struct hero` or is there some other aspect you're  referring to?

Reply:
I am thinking of the traits used by p2p, which is a list of key-value pairs. maybe introducing a standardized object for the traits.

Reply:
I think this is a great opportunity for the community to dictate that this is a feature that they'd like. The tokenv2 model is intended to be opinionless and simple thus enabling standardization to bake in over time.